7$: Report error on BOM before #!

BOM: Byte Order Mark
#!: Shebang

Linux (and I assume macOS and other POSIX-like platforms) only interpret shebangs if the first byte in the file is `#` and the second byte in the file is `!` (I think; please confirm). This means that the presence of a BOM (UTF-8-encoded) will cause the shebang characters to not be interpreted by the operating system. This is almost definitely a mistake.

Report an error if this mistake occurs.

Is this the correct place to look in for this issue? - https://github.com/quick-lint/quick-lint-js/blob/8ab6bed96f16678b686395643728e580e6e8d0e0/src/lex.cpp#L543

Right now for this program (with BOM character before #) - 
```javascript
#!/usr/bin/env node
let x = 3;
console.log(x);
```
quick-lint-js report the following errors and warnings - 
```shell
main.js:1:4: error: unexpected '#' [E052]
main.js:1:18: error: missing semicolon after statement [E027]
main.js:1:15: warning: use of undeclared variable: env [E057]
main.js:1:19: warning: use of undeclared variable: node [E057]
```
What error should is expected to be reported?

> Is this the correct place to look in for this issue? -
> 
> https://github.com/quick-lint/quick-lint-js/blob/8ab6bed96f16678b686395643728e580e6e8d0e0/src/lex.cpp#L543

I don't think so. When we execute that piece of code, we skipped the BOM already.

> Right now for this program (with BOM character before #) -
> 
> ```js
> #!/usr/bin/env node
> let x = 3;
> console.log(x);
> ```
> 
> quick-lint-js report the following errors and warnings -
> 
> ```shell
> main.js:1:4: error: unexpected '#' [E052]
> main.js:1:18: error: missing semicolon after statement [E027]
> main.js:1:15: warning: use of undeclared variable: env [E057]
> main.js:1:19: warning: use of undeclared variable: node [E057]
> ```
> 
> What error should is expected to be reported?

`error: Unicode byte order mark (BOM) cannot appear before #! at beginning of script`

--- a/docs/errors/E095.md
+++ b/docs/errors/E095.md
@@ -1,0 +1,13 @@
+# E095: unicode byte order mark (BOM) cannot appear before #! at beginning of script
+
+A script cannot have unicode byte order mark (BOM) before `#!`
+
+    #!/usr/bin/env node
+    let x = 3;
+    console.log(x);
+
+To fix this error, remove the unicode BOM before `#!`
+
+    #!/usr/bin/env node
+    let x = 3;
+    console.log(x);

--- a/docs/errors/E095.md
+++ b/docs/errors/E095.md
@@ -1,12 +1,12 @@
-# E095: unicode byte order mark (BOM) cannot appear before #! at beginning of script
+# E095: Unicode byte order mark (BOM) cannot appear before #! at beginning of script
 
-A script cannot have unicode byte order mark (BOM) before `#!`
+A script cannot have Unicode byte order mark (BOM) before `#!`
 
     #!/usr/bin/env node
     let x = 3;
     console.log(x);
 
-To fix this error, remove the unicode BOM before `#!`
+To fix this error, remove the Unicode BOM before `#!`
 
     #!/usr/bin/env node
     let x = 3;

--- a/src/lex.cpp
+++ b/src/lex.cpp
@@ -175,7 +175,7 @@ void lexer::parse_bom_before_shebang() {
     if (static_cast<unsigned char>(input[0]) == '#' &&
         static_cast<unsigned char>(input[1]) == '!') {
       this->error_reporter_->report(error_unexpected_bom_before_shebang{
-          source_code_span(&this->input_[0], &this->input_[2])});
+          source_code_span(&this->input_[0], &this->input_[3])});
       input += 2;
       this->skip_line_comment_body();
     } else {

--- a/src/lex.cpp
+++ b/src/lex.cpp
@@ -162,7 +162,26 @@ lexer::lexer(padded_string_view input, error_reporter* error_reporter) noexcept
       error_reporter_(error_reporter),
       original_input_(input) {
   this->last_token_.end = this->input_;
+  this->parse_bom_before_shebang();
   this->parse_current_token();
+}
+
+void lexer::parse_bom_before_shebang() {
+  const char8* input = this->input_;
+  if (static_cast<unsigned char>(input[0]) == 0xef &&
+      static_cast<unsigned char>(input[1]) == 0xbb &&
+      static_cast<unsigned char>(input[2]) == 0xbf) {
+    input += 3;
+    if (static_cast<unsigned char>(input[0]) == '#' &&
+        static_cast<unsigned char>(input[1]) == '!') {
+      this->error_reporter_->report(error_unexpected_bom_before_shebang{
+          source_code_span(&this->input_[0], &this->input_[2])});
+      input += 2;
+      this->skip_line_comment_body();
+    } else {
+      this->input_ = input;
+    }
+  }
 }
 
 void lexer::parse_current_token() {

--- a/src/quick-lint-js/error.h
+++ b/src/quick-lint-js/error.h
@@ -1050,6 +1050,13 @@
       .error(QLJS_TRANSLATABLE("unexpected '#'"), where))                      \
                                                                                \
   QLJS_ERROR_TYPE(                                                             \
+      error_unexpected_bom_before_shebang, "E095",                             \
+      { source_code_span where; },                                             \
+      .error(QLJS_TRANSLATABLE(                                                \
+                 "unicode byte order mark (BOM) cannot appear before #! "      \
+                 "at beginning of script"), where))                            \
+                                                                               \
+  QLJS_ERROR_TYPE(                                                             \
       error_unexpected_identifier, "E053", { source_code_span where; },        \
       .error(QLJS_TRANSLATABLE("unexpected identifier"), where))               \
                                                                                \

--- a/src/quick-lint-js/error.h
+++ b/src/quick-lint-js/error.h
@@ -1050,12 +1050,11 @@
       .error(QLJS_TRANSLATABLE("unexpected '#'"), where))                      \
                                                                                \
   QLJS_ERROR_TYPE(                                                             \
-      error_unexpected_bom_before_shebang, "E095",                             \
-      { source_code_span where; },                                             \
+      error_unexpected_bom_before_shebang, "E095", { source_code_span bom; },  \
       .error(QLJS_TRANSLATABLE(                                                \
                  "unicode byte order mark (BOM) cannot appear before #! "      \
                  "at beginning of script"),                                    \
-             where))                                                           \
+             bom))                                                             \
                                                                                \
   QLJS_ERROR_TYPE(                                                             \
       error_unexpected_identifier, "E053", { source_code_span where; },        \

--- a/src/quick-lint-js/error.h
+++ b/src/quick-lint-js/error.h
@@ -1054,7 +1054,8 @@
       { source_code_span where; },                                             \
       .error(QLJS_TRANSLATABLE(                                                \
                  "unicode byte order mark (BOM) cannot appear before #! "      \
-                 "at beginning of script"), where))                            \
+                 "at beginning of script"),                                    \
+             where))                                                           \
                                                                                \
   QLJS_ERROR_TYPE(                                                             \
       error_unexpected_identifier, "E053", { source_code_span where; },        \

--- a/src/quick-lint-js/lex.h
+++ b/src/quick-lint-js/lex.h
@@ -199,6 +199,7 @@ class lexer {
     escape_sequence_list* escape_sequences;
   };
 
+  void parse_bom_before_shebang();
   void parse_current_token();
 
   const char8* parse_string_literal() noexcept;

--- a/test/test-lex.cpp
+++ b/test/test-lex.cpp
@@ -1876,18 +1876,6 @@ TEST_F(test_lex, lex_not_shebang) {
                               offsets_matcher(&input, 2, 3))));
   }
 
-  // BOM must not appear before '#!'.
-  {
-    error_collector v;
-    padded_string input(u8"\ufeff#!notashebang\n"_sv);
-    lexer l(&input, &v);
-    EXPECT_EQ(l.peek().type, token_type::end_of_file) << "# should be skipped";
-
-    EXPECT_THAT(v.errors, ElementsAre(ERROR_TYPE_FIELD(
-                              error_unexpected_bom_before_shebang, where,
-                              offsets_matcher(&input, 0, 2))));
-  }
-
   {
     error_collector v;
     padded_string input(u8"#\\u{21}\n"_sv);
@@ -1900,6 +1888,20 @@ TEST_F(test_lex, lex_not_shebang) {
         ElementsAre(ERROR_TYPE_FIELD(
             error_escaped_character_disallowed_in_identifiers, escape_sequence,
             offsets_matcher(&input, strlen(u8"#"), u8"\\u{21}"))));
+  }
+}
+
+TEST_F(test_lex, lex_unexpected_bom_before_shebang) {
+  // BOM must not appear before '#!'.
+  {
+    error_collector v;
+    padded_string input(u8"\ufeff#!notashebang\n"_sv);
+    lexer l(&input, &v);
+    EXPECT_EQ(l.peek().type, token_type::end_of_file) << "# should be skipped";
+
+    EXPECT_THAT(v.errors, ElementsAre(ERROR_TYPE_FIELD(
+                              error_unexpected_bom_before_shebang, bom,
+                              offsets_matcher(&input, 0, u8"\ufeff"))));
   }
 }
 

--- a/test/test-lex.cpp
+++ b/test/test-lex.cpp
@@ -1876,6 +1876,18 @@ TEST_F(test_lex, lex_not_shebang) {
                               offsets_matcher(&input, 2, 3))));
   }
 
+  // BOM must not appear before '#!'.
+  {
+    error_collector v;
+    padded_string input(u8"\ufeff#!notashebang\n"_sv);
+    lexer l(&input, &v);
+    EXPECT_EQ(l.peek().type, token_type::end_of_file) << "# should be skipped";
+
+    EXPECT_THAT(v.errors, ElementsAre(ERROR_TYPE_FIELD(
+                              error_unexpected_bom_before_shebang, where,
+                              offsets_matcher(&input, 0, 2))));
+  }
+
   {
     error_collector v;
     padded_string input(u8"#\\u{21}\n"_sv);

